interop-reactiveStreams/shared/src/main/scala/scalaz/zio/interop/reactiveStreams/SinkSubscriber.scala

```diff
@@ -0,0 +1,65 @@
+package scalaz.zio.interop.reactiveStreams
+
+import org.reactivestreams.{ Subscriber, Subscription }
+import scalaz.zio.stream.{ Sink, Stream }
+import scalaz.zio.{ Promise, Queue, Runtime, Task, ZIO }
+
+class SinkSubscriber[A, B](
+  runtime: Runtime[_],
+  q: Queue[A],
+  p: Promise[Throwable, B],
+  qSize: Int
+) extends Subscriber[A] {
+
+  // all signals in reactive streams are serialized, so we don't need any synchronization
+  private var subscriptionOpt: Option[Subscription] = None
+  private var signalledDemand                       = qSize
+
+  override def onSubscribe(subscription: Subscription): Unit =
+    if (subscription == null) {
+      throw new NullPointerException("s was null in onSubscribe")
+    } else if (subscriptionOpt.isDefined) {
+      subscription.cancel()
+    } else {
+      subscriptionOpt = Some(subscription)
+      runtime.unsafeRunAsync(q.awaitShutdown *> Task(subscription.cancel()))(_ => ())
+      // see reactive streams rule 3.17. We do not track demand beyond Long.MaxValue
+      subscription.request(qSize.toLong)
+    }
+
+  override def onNext(t: A): Unit =
+    if (t == null) {
+      throw new NullPointerException("t was null in onNext")
+    } else {
+      runtime.unsafeRun(q.offer(t))
+      if (signalledDemand > 1) {
+        signalledDemand -= 1
+      } else {
+        subscriptionOpt.foreach(_.request(qSize.toLong))
```

Comment:
So if I am not mistaken, at this point you have no guarantee that the queue has been emptied and it is safe to issue more demand, as next calls to `onNext` may block on the `offer`.

Do we need to link dequeues from the queue to the demand?

Reply:
Right. I should take a break and really work on this later.

Reply:
No rush! Thank you for spending the time on this.